### DISCUSSION TITLE
Add missing bond if obvious

### DISCUSSION THREAD

**Discussion by dln22** [rdkit/rdkit#4259] [Q&A] [answered]
I am looking for some help. My task is to read a bunch of xyz files representing molecule structure along the bond scan. So one can imagine that the xyz files will contain fragments at larger and larger distance. I have figured how to detect fragments and create mol objects for them in rdkit. There is also a command to combine fragments into a single molecule, however the bond between them is not automatically created. I could simply specify the bond to be inserted, but am wondering if, for simple cases where we have two radical fragments , rdkit could somehow automatically detect which bond is missing?

Here is my code snapshot:

```python
def getMolFragsInchi(inputMol, molFormat):
    """Given molecule file in e.g. xyz format, detects if it contains fragments
         and returns their inchi strings.""""
    inchi = obConvert(molFormat, 'inchi', inputMol) # openbabel wrapper
    mol_inchi = rdkit.Chem.inchi.MolFromInchi(inchi)
    mol_frags = rdkit.Chem.GetMolFrags(mol_inchi,asMols=True)
    inchi_list = []
    for m in mol_frags:
        inchi_list.append(rdkit.Chem.inchi.MolToInchi(m))
    return inchi_list

def molFragsToInchi(inputMol, molFormat):
    """Given two inchi fragment strings, create their corresponding mol
         objects and combine them into a single molecule.""""
    inchi_list = getMolFragsInchi(inputMol, molFormat)

    m1 = rdkit.Chem.inchi.MolFromInchi(inchi_list[0])
    m2 = rdkit.Chem.inchi.MolFromInchi(inchi_list[1])

    #m12 still represent two disjoint fragments
    m12 = rdkit.Chem.CombineMols(m1,m2)

    return rdkit.Chem.inchi.MolToInchi(m12)
````

**Comment by ptosco** [comment, marked as answer]
@dln22 after using `CombineMols` you can cast your mol to a `Chem.RWMol` and use `AddBond` between the relevant atom indices. If the disjoint fragments have explicit Hs and have open valences, after creating the bond you should be able to call `Chem.SanitizeMol` on the resulting joined molecule without issues.
`Chem.MolFromPDBBlock` has logic to infer bonds based on proximity, not open valences.
If you do not know the indices but have open valences you could loop on heavy atoms and find those whose valence is < than the atom's default valence.

**Reply by dln22** [reply]
Paolo, once again many thanks for a great answer. Any chance you could please provide a minimal code example on how to find atoms default valences? I am not super experienced with rdkit yet. So far I have tried to use the `GetExplicitValence` and `GetImplicitValence` functions. I found that the `GetExplicitValence` correctly gives the current heavy atom valences in a molecule, e.g. for CH3 radical it gives valence 3 for the carbon atom. However, the `GetImplicitValence` function for the same molecule and the same carbon atom gives the valence 0 (so it probably does something else that I think). I also tried to create a C atom in isolation and run the `GetExplicitValence` on it, but this results in an exception that atoms valences are only defined if atoms are part of a molecule.

**Reply by ptosco** [reply]
@dln22 it would be great if you could attach a zip file with some examples of your fragments pairs, so I have something to work on.

**Reply by dln22** [reply]
Hi Paolo, apologies for the slight delay in providing example scan points. Please see them attached below. Most of our bond scans produce two fragments with a single radical site on each. However, we also have bond scans on di-radical parent molecules that produce fragments with four radical sites. In such cases, it is not possible to find which fragments atoms should be bonded to create the parent molecule by examining only the atoms valences. One rule that might work here, I think, is to pick the fragments atoms based on their valences and distance from each other (if multiple bonding options exist, we should pick the atoms that are closest).

In the attached zip archive you will find 7 xyz files containing two fragments at some distance. The name of each example file also contains the line numbers e.g. L5_L10, here line 5 and line 10 in xyz file that has atoms that should be bonded to create the parent molecule. Many thanks for help!

[scanPointsExamples.zip](https://github.com/rdkit/rdkit/files/6709205/scanPointsExamples.zip)

**Reply by dln22** [reply]
hello,

I have been trying to come up with some solution to this problem, but with no luck so far. Here is what I have tried (currently assuming only the most simple case, that the fragments would have a single radical site each):


```python

def getMolFrags(inputMol, molFormat):
    """
    Given xyz file, convert it to inchi and then to a list of rdkit mol objects.
    As an example, for an xyz file containing ch3 and ch2oh fragments at
    some distance the inchi string returned from the obConvert will be:
    'InChI=1S/CH3O.CH3/c1-2;/h2H,1H2;1H3'
    the inchi string will further be used to create rdkit mol object and later
    split into two mol fragment objects.         
    """
         
    inchi = obConvert(molFormat, 'inchi', inputMol)
    mol_inchi = rdkit.Chem.MolFromInchi(inchi)

    mol_frags = rdkit.Chem.GetMolFrags(mol_inchi,asMols=True)
    for mol in mol_frags:
        rdkit.Chem.SanitizeMol(mol)
    return mol_frags

def findRadicalSite(mol):
    """
    Returns the first atom index with a non-zero number of radical sites.
    """
    atomid = -1
    for atom in mol.GetAtoms():
        if atom.GetNumRadicalElectrons() > 0:
            atomid = atom.GetAtomMapNum()
            break
    return atomid

def molFragsToMol(inputMol, molFormat):
    """
    Given xyz file containing two molecular fragments, 
    recombines them back into the parent molecule.
    """
    mol_frags = getMolFrags(inputMol, molFormat)  
    
    atom1id = findRadicalSite(mol_frags[0])
    atom2id = findRadicalSite(mol_frags[1]) + mol_frags[0].GetNumAtoms()
    
    m12 = rdkit.Chem.CombineMols(mol_frags[0],mol_frags[1])
    m12 = rdkit.Chem.RWMol(m12)
    m12.AddBond(atom1id,atom2id)
    rdkit.Chem.SanitizeMol(m12)
    return m12


# when I finally put everything together:
mol = molFragsToMol(ch3_ch2oh_file,'xyz')
````

Here is the final mol file. As can be seen the radical sites are still there.
![image](https://user-images.githubusercontent.com/17905932/123811410-1cfa1e00-d8eb-11eb-8782-182b2a78a1c6.png)

I also tried to add hydrogens to the final mol object:
```python
    mol=rdkit.Chem.rdmolops.AddHs(mol)
```
However, this also didn't eliminate the radical sites.
![image](https://user-images.githubusercontent.com/17905932/123811613-4fa41680-d8eb-11eb-9fac-3042fb7f4f03.png)

Another problem is that if I try to convert the final mol object back to inchi via:
```python
    rdkit.Chem.inchi.MolToInchi(mol)
```
it returns an empty string, so something is wrong. I would very much appreciate any help here. Being able to convert the final mol structure back to inchi is very important here as it would allow me to find the parent molecule.

**Reply by ptosco** [reply]
@dln22 Your code works with just a couple of adjustments:
```
def findRadicalSite(mol):
    """
    Returns the first atom index with a non-zero number of radical sites.
    """
    atomid = -1
    for atom in mol.GetAtoms():
        if atom.GetNumRadicalElectrons() > 0:
            atomid = atom.GetIdx()
            break
    return atomid

def molFragsToMol(inputMol, molFormat):
    """
    Given xyz file containing two molecular fragments, 
    recombines them back into the parent molecule.
    """
    mol_frags = getMolFrags(inputMol, molFormat)  
    
    atom1id = findRadicalSite(mol_frags[0])
    atom2id = findRadicalSite(mol_frags[1]) + mol_frags[0].GetNumAtoms()
    
    m12 = rdkit.Chem.CombineMols(mol_frags[0], mol_frags[1])
    m12 = rdkit.Chem.RWMol(m12)
    m12.AddBond(atom1id, atom2id, Chem.BondType.SINGLE)
    for atomId in (atom1id, atom2id):
        atom = m12.GetAtomWithIdx(atomId)
        atom.SetNumRadicalElectrons(0)
    rdkit.Chem.SanitizeMol(m12)
    return m12
```

**Reply by dln22** [reply]
Perfect, many thanks for the fix!
